File: impl/go/internal/tool/bash.go
// bash — run a terminal command inside the caged sandbox (Go parity of bash.ts).
package tool

import "fmt"

const defaultTimeoutMS = 120_000 // 2 minutes

// BashTool wraps backend.Exec (already caged by a Job Object + jailed workdir).
var BashTool = Def{
	ID: "bash",
	Description: "Run a terminal command inside the project sandbox (git, npm, npx, remotion, ls, etc.). " +
		"The sandbox is resource-capped and jailed to the project workdir. Prefer the read/write/edit/grep/glob tools for file work; " +
		"use bash for builds, installs, and running commands. Default timeout 120s.",
	Parameters: []Param{
		{Name: "command", Type: "string", Description: "The command to execute", Required: true},
		{Name: "description", Type: "string", Description: "Clear, concise description of what this command does (5-10 words)", Required: true},
		{Name: "timeout", Type: "number", Description: "Optional timeout in milliseconds (default 120000)"},
	},
	Validate: func(args map[string]any) (map[string]any, error) {
		command, err := reqString(args, "command")
		if err != nil {
			return nil, err
		}
		description, err := reqString(args, "description")
		if err != nil {
			return nil, err
		}
		timeout, _, err := optNumber(args, "timeout")
		if err != nil {
			return nil, err
		}
		return map[string]any{"command": command, "description": description, "timeout": timeout}, nil
	},
	Execute: func(args map[string]any, ctx *Context) (Result, error) {
		command := args["command"].(string)
		description := args["description"].(string)
		timeoutMS := defaultTimeoutMS
		if t, ok := args["timeout"].(float64); ok && t > 0 {
			timeoutMS = int(t)
		}
		ctx.LogStep("command", description, StepExtra{Kind: "command", Command: command})

		r, err := ctx.Backend.Exec(ctx.SandboxID, command, "", nil, float64(timeoutMS)/1000.0)
		if err != nil {
			return Result{Title: description, Output: "Command failed to start: " + err.Error(), Metadata: map[string]any{"error": "exec_failed"}}, nil
		}
		combined := r.Stdout
		if r.Stderr != "" {
			if combined != "" {
				combined += "\n"
			}
			combined += r.Stderr
		}
		body := combined
		if body == "" {
			body = "(no output)"
		}
		if r.TimedOut {
			body += fmt.Sprintf("\n\n<shell_metadata>\nCommand terminated after exceeding the %d ms timeout. "+
				"If it needs longer and is not waiting for interactive input, retry with a larger timeout.\n</shell_metadata>", timeoutMS)
		} else if r.KilledByCap {
			body += "\n\n<shell_metadata>\nProcess killed by the sandbox resource cap (RAM/CPU). Reduce memory use or split the work.\n</shell_metadata>"
		}
		out, truncated, outputPath := truncateOutput(ctx, "bash", body, 30*1024)
		exit := r.ExitCode
		ctx.LogStep("command", fmt.Sprintf("exit %d", exit), StepExtra{Kind: "command_output", ExitCode: &exit, Output: combined})
		return Result{
			Title:      description,
			Output:     out,
			Truncated:  truncated,
			OutputPath: outputPath,
			// A nonzero exit is NOT a soft failure (the model often expects it);
			// the model reads the exit line in the output.
			Metadata: map[string]any{"exit": r.ExitCode, "timedOut": r.TimedOut, "killedByCap": r.KilledByCap},
		}, nil
	},
}

File: impl/ts/src/tool/bash.ts
// bash — run a terminal command inside the project's caged sandbox.
// See docs/harness-design.md §5.3. Wraps backend.exec (already caged by a Job
// Object: RAM/CPU/process caps + jailed workdir), so unlike opencode there is no
// tree-sitter parse or per-command permission prompt. Output is tail-limited.

import type { ToolDef, ToolContext, ToolResult } from "./types.js";
import { reqString, optNumber, optString } from "./validate.js";
import { truncateOutput } from "./truncate.js";

const DEFAULT_TIMEOUT_MS = 120_000; // 2 minutes, matching opencode's bash default

interface Args {
  command: string;
  description: string;
  timeout?: number; // milliseconds
}

export const BashTool: ToolDef<Args> = {
  id: "bash",
  description:
    "Run a terminal command inside the project sandbox (git, npm, npx, remotion, ls, etc.). The sandbox is resource-capped and jailed to the project workdir. Prefer the read/write/edit/grep/glob tools for file work; use bash for builds, installs, and running commands. Default timeout 120s.",
  parameters: {
    command: { type: "string", description: "The command to execute", required: true },
    description: { type: "string", description: "Optional short description of what this command does (for the transcript)" },
    timeout: { type: "number", description: "Optional timeout in milliseconds (default 120000)" },
  },
  validate(args) {
    const command = reqString(args, "command");
    // description is for the transcript only — don't block the tool if a model
    // omits it; fall back to a truncated form of the command itself.
    const description = optString(args, "description") || command.slice(0, 60);
    return { command, description, timeout: optNumber(args, "timeout") };
  },
  async execute(args, ctx: ToolContext): Promise<ToolResult> {
    const timeoutMs = args.timeout && args.timeout > 0 ? args.timeout : DEFAULT_TIMEOUT_MS;
    ctx.log("command", args.description, { kind: "command", command: args.command });

    const r = await ctx.backend.exec(ctx.sandboxId, args.command, {
      timeoutS: Math.ceil(timeoutMs / 1000),
    });

    const combined = [r.stdout, r.stderr].filter((s) => s && s.trim() !== "").join("\n");
    let body = combined.trim() === "" ? "(no output)" : combined;
    if (r.timedOut) {
      body +=
        `\n\n<shell_metadata>\nCommand terminated after exceeding the ${timeoutMs} ms timeout. ` +
        `If it needs longer and is not waiting for interactive input, retry with a larger timeout.\n</shell_metadata>`;
    } else if (r.killedByCap) {
      body += `\n\n<shell_metadata>\nProcess killed by the sandbox resource cap (RAM/CPU). Reduce memory use or split the work.\n</shell_metadata>`;
    }

    const shaped = truncateOutput(ctx, "bash", body, 30 * 1024);
    ctx.log("command", `exit ${r.exitCode}`, {
      kind: "command_output",
      exitCode: r.exitCode,
      output: combined,
    });
    return {
      title: args.description,
      output: shaped.output,
      truncated: shaped.truncated,
      outputPath: shaped.outputPath,
      // exit code in metadata is informational; a nonzero exit is NOT a soft
      // failure (the model often runs commands expecting nonzero), so we do not
      // set metadata.error here — the model reads the exit line in the output.
      metadata: { exit: r.exitCode, timedOut: r.timedOut, killedByCap: r.killedByCap },
    };
  },
};
